### PR TITLE
Resolve UB found by miri

### LINKED ISSUE #41: Resolve UB found by miri
Miri reports plenty of UB in Tendril crate. It can be tested checked using `cargo miri test` (on nightly).

### LINKED ISSUE #44: Use an union to represent inline allocation
This is necessary for https://github.com/servo/tendril/issues/41. Replace `len` and `aux` fields in `Tendril` struct with a `(Unsafe)Cell<union>` field. The union should be able to represent both inline representation as well as heap representation. This may allow removing `#[repr(C)]` too.

### PR DESCRIPTION
**Pull request by xfix**
Fixes #41. Fixes #44.

### CODE REVIEW

**Comment by xfix** [comment]
Hm, compatibility with C API may be tricky. `Header` cannot be `packed` because `Atomicity` takes methods by `&self` (you cannot create a valid reference to a packed struct).

**Comment by xfix** [comment]
`AtomicUsize` takes methods by reference (see https://doc.rust-lang.org/std/sync/atomic/struct.AtomicUsize.html), which means that atomics cannot be packed. I think changes in C API will be necessary for this to keep working.

**Comment by xfix** [comment]
This change breaks C ABI compatibility (requiring new `tendril.h`), keep that in mind. I don't think this soundness change can be applied without doing so - unaligned atomic accesses are not legal.

**Comment by xfix** [comment]
Unless... only `Header<Atomic>` would be incompatible, but not `Header<NonAtomic>`. I think that may be possible to support.

**Comment by xfix** [comment]
Okay, this introduces a breaking change where `Tendril<_, Atomic>` is no longer C safe. `Tendril<_, NonAtomic>` is still fine. The C API doesn't allow to create `Tendril<_, Atomic>`, so this should be acceptable, I hope.

**Comment by xfix** [comment]
Wait, no, `Header` cannot be `packed`, period. This is because we use `let shared = (ptr & 1) == 1;` for determining if memory is shared, but `packed` memory has alignment of 1. I suppose I can use `align(4)` to force workable alignment?

**Comment by xfix** [comment]
Okay, I think that's it. Miri found legitimate bugs in the library: `packed` struct where the program depended on particular alignment and unaligned atomic writes. An alignment aware allocator (not default `malloc`) would have broken the library.

**Comment by SimonSapin** [comment]
I’d be inclined to drop the C API. I don’t know of anyone using it, and it’s not really maintained.

File: src/tendril.rs [hunk 1]
```diff
@@ -68,30 +74,56 @@ pub unsafe trait Atomicity: 'static {
 /// and so doesn't typically need to be written.
 ///
 /// This is akin to using `Rc` for reference counting.
-pub struct NonAtomic(Cell<usize>);
+#[repr(C, align(4))]
+pub struct NonAtomic(Cell<NonAtomicInternals>);
```
Comment (SimonSapin) [inline src/tendril.rs]: Wouldn’t a struct of separate cells be more convenient that a cell of struct?

Comment (xfix) [reply src/tendril.rs]: Probably, yes. This is a leftover from when the the entire structure was packed, but only the reference counter needs to be packed now.

**Comment by xfix** [comment]
As for C API, I think it should be removed with 0.5.0, but it's not necessary to remove it for this change.

**Comment by xfix** [comment]
Updated to run thread tests on miri, now that Miri supports threads.

File: src/tendril.rs [hunk 2]
```diff
@@ -1162,6 +1180,22 @@ where
...
+
+    unsafe fn set_len(&mut self, len: u32) {
+        (*self.buf.get()).heap.len = len;
+    }
+
+    unsafe fn aux(&self) -> u32 {
+        (*self.buf.get()).heap.aux
+    }
+
+    unsafe fn set_aux(&self, aux: u32) {
```
Comment (jdm) [inline src/tendril.rs:1196]: Should this be `&mut self` like set_len?

Comment (xfix) [reply src/tendril.rs:1196]: No because it's used by `make_buf_shared` which is used by public `unsafe_subtendril`. This isn't a public method anyway.

**Comment by jdm** [comment]
@bors-servo r+
Thanks!

**Pull request merged**